src/wp-includes/class-wp-token-map.php

```diff
@@ -0,0 +1,740 @@
+ * may be too small for a key length of 2. Perhaps 1 would be more appropriate. It's
+ * best to experiment and determine empirically which values are appropriate.
+ *
+ * ## Generate Pre-Computed Source Code.
+ *
+ * Since the `WP_Token_Map` is designed for relatively static lookups, it can be
+ * advantageous to precompute the values and instantiate a table that has already
+ * sorted and grouped the tokens and built the lookup strings.
+ *
+ * This can be done with `WP_Token_Map::precomputed_php_source_table()`.
+ *
+ * Note that if there is a leading character that all tokens need, such as `&` for
+ * HTML named character references, it can be beneficial to exclude this from the
+ * token map. Instead, find occurrences of the leading character and then use the
+ * token map to see if the following characters complete the token.
+ *
+ * Example:
+ *
+ *     $map = WP_Token_Map::from_array( array( 'simple_smile:' => '🙂', 'sob:' => '😭', 'soba:' => '🍜' ) );
+ *     echo $map->precomputed_php_source_table();
+ *     // Output
+ *     WP_Token_Map::from_precomputed_table(
+ *         2,
+ *         "si\x00so\x00",
+ *         array(
+ *                 // simple_smile:[🙂].
+ *                 "\x0bmple_smile:\x04🙂",
+ *                 // soba:[🍜] sob:[😭].
+ *                 "\x03ba:\x04🍜\x02b:\x04😭",
+ *         ),
+ *         "",
+ *         array()
+ *     );
+ *
+ * This precomputed value can be stored directly in source code and will skip the
+ * startup cost of generating the lookup strings. See `$html5_named_character_entities`.
+ *
+ * ## Future Direction.
+ *
+ * It may be viable to dynamically increase the length limits such that there's no need to impose them.
+ * The limit appears because of the packing structure, which indicates how many bytes each segment of
+ * text in the lookup tables spans. If, however, care were taken to track the longest word length, then
+ * the packing structure could change its representation to allow for that. Each additional byte storing
+ * length, however, increases the memory overhead and lookup runtime.
+ *
+ * An alternative approach could be to borrow the UTF-8 variable-length encoding and store lengths of less
+ * than 127 as a single byte with the high bit unset, storing longer lengths as the combination of
+ * continuation bytes.
+ *
+ * Since it has not been shown during the development of this class that longer strings are required, this
+ * update is deferred until such a need is clear.
+ *
+ * @since 6.6.0
+ */
+class WP_Token_Map {
+	/**
+	 * Maximum length for each key and each transformed value in the table (in bytes).
+	 *
+	 * @since 6.6.0
+	 */
+	const MAX_LENGTH = 256;
+
+	/**
+	 * How many bytes of each key are used to form a group key for lookup.
+	 * This also determines whether a word is considered short or long.
+	 *
+	 * @since 6.6.0
+	 *
+	 * @var int
+	 */
+	private $key_length = 2;
+
+	/**
+	 * Stores an optimized form of the word set, where words are grouped
```

Comment:
I'd like to see the explicit rule for large vs. small words in their docs:

> …where large words are strings whose strlen greater than is key_length

(that's almost certainly wrong, just an example)

Reply:
added class-level description in [31c57b0](https://github.com/WordPress/wordpress-develop/pull/5373/commits/31c57b0871a193fdf18a6ae870cb32b0750799a2)